### ISSUE TITLE
Microsoft.Data.SqlClient.SqlException: 'Argument data type NULL is invalid for argument 2 of JSON_VALUE function.'

### ISSUE THREAD

**Issue by ArkadiuszMakosa** [dotnet/efcore#32952]
Today I experienced weird behaviour of EF Core. 
Long story short we have **_User_** Entity with **_UserAdGroupSynchronizationErrors_** Entity with optional one to one relation. 

Because of fact, it is optional in mapping I checking if _UserAdGroupSynchronizationErrors_ is null:

```C#
private static TimeSpan[] TimeSpans = [TimeSpan.FromDays(1), TimeSpan.FromDays(2), TimeSpan.FromDays(7), TimeSpan.FromDays(10 * 365)];

public override Expression<Func<User, UserDto>> GetMapping()
{
	return (e) => new UserDto
	{
		NextReRunDate = e.UserAdGroupSynchronizationErrors != null ?
			e.UserAdGroupSynchronizationErrors.LastTimeTry.Add(TimeSpans[e.UserAdGroupSynchronizationErrors.RetryNumber]) : null
	};
}
```

Such mapping cause the exception **Microsoft.Data.SqlClient.SqlException: 'Argument data type NULL is invalid for argument 2 of JSON_VALUE function.'**

If we take a look at QueryString EFCore generate for us we can see, that regardless of nullcheck SQL try to Cast 

```SQL
CAST(JSON_VALUE(@__TimeSpans_1, '$[' + CAST([u0].[RetryNumber] AS nvarchar(max)) + ']') AS time)
```

Let me put whole SQL query for context:

```SQL
DECLARE @__TimeSpans_1 nvarchar(4000) = N'["1:0:00:00","2:0:00:00","7:0:00:00","3650:0:00:00"]';
DECLARE @__AllowedIds_0 nvarchar(4000) = N'[338,355,387,972,1351,2168,2419]';
DECLARE @__p_2 int = 0;
DECLARE @__p_3 int = 20;

SELECT CASE
    WHEN [u0].[Id] IS NOT NULL THEN CAST(1 AS bit)
    ELSE CAST(0 AS bit)
END, [u0].[LastTimeTry], CAST(JSON_VALUE(@__TimeSpans_1, '$[' + CAST([u0].[RetryNumber] AS nvarchar(max)) + ']') AS time)
FROM (
    SELECT [u].[Id]
    FROM [Identity].[Users] AS [u]
    WHERE [u].[Id] IN (
        SELECT [a].[value]
        FROM OPENJSON(@__AllowedIds_0) WITH ([value] bigint '$') AS [a]
    )
    ORDER BY [u].[Id]
    OFFSET @__p_2 ROWS FETCH NEXT @__p_3 ROWS ONLY
) AS [t]
LEFT JOIN [Identity].[UserAdGroupSynchronizatioinErrors] AS [u0] ON [t].[Id] = [u0].[UserId]
ORDER BY [t].[Id]
```

```C#
public class User 
{
	public long Id { get; set; }

	public UserAdGroupSynchronizationError UserAdGroupSynchronizationErrors { get; set; }

}

public class UserAdGroupSynchronizationError : IIdentifiable
{
	public long Id { get; set; }

	public DateTime LastTimeTry { get; set; }

	public int RetryNumber { get; set; }
}
```

### SOLUTION:

As a Quick fix I can use such mapping:

```C#
private static TimeSpan[] TimeSpans = [TimeSpan.FromDays(1), TimeSpan.FromDays(2), TimeSpan.FromDays(7), TimeSpan.FromDays(10 * 365)];

public override Expression<Func<User, UserDto>> GetMapping()
{
	return (e) => new UserDto
	{
		NextReRunDate = e.UserAdGroupSynchronizationErrors != null ?
			e.UserAdGroupSynchronizationErrors.LastTimeTry.Add(TimeSpans[e.UserAdGroupSynchronizationErrors != null ? e.UserAdGroupSynchronizationErrors.RetryNumber : 0]) : null
	};
}
```

After such change we will get: 

```SQL
DECLARE @__TimeSpans_1 nvarchar(4000) = N'["1:0:00:00","2:0:00:00","7:0:00:00","3650:0:00:00"]';
DECLARE @__AllowedIds_0 nvarchar(4000) = N'[338,355,387,972,1351,2168,2419]';

SELECT CASE
    WHEN [u0].[Id] IS NOT NULL THEN CAST(1 AS bit)
    ELSE CAST(0 AS bit)
END, [u0].[LastTimeTry], CAST(JSON_VALUE(@__TimeSpans_1, '$[' + CAST(CASE
    WHEN NOT ([u0].[Id] = NULL) THEN [u0].[RetryNumber]
    ELSE 0
END AS nvarchar(max)) + ']') AS time)
FROM [Identity].[Users] AS [u]
LEFT JOIN [Identity].[UserAdGroupSynchronizatioinErrors] AS [u0] ON [u].[Id] = [u0].[UserId]
WHERE [u].[Id] IN (
    SELECT [a].[value]
    FROM OPENJSON(@__AllowedIds_0) WITH ([value] bigint '$') AS [a]
)
ORDER BY [u].[Id]
```

with response:

		
| (No column name) | LastTimeTry | (No column name) |
|------------------|-------------|------------------|
| 0                | NULL        | 01:00:00.0000000 |
| 0                | NULL        | 01:00:00.0000000 |
| 0                | NULL        | 01:00:00.0000000 |
| 0                | NULL        | 01:00:00.0000000 |
| 0                | NULL        | 01:00:00.0000000 |
| 0                | NULL        | 01:00:00.0000000 |
| 0                | NULL        | 01:00:00.0000000 |

# To sum up
Double null check with default value works fine, but this is not code I would expect to see in my project codebase :)

### Include provider and version information
EF Core version: 8.0.0
EntityFrameworkCore.SqlServer: 8.0.0
Target framework: .Net8
MS SQL Server: 16.0.4105.2
Operating system: Windows

**Comment by maumar** [comment]
@ArkadiuszMakosa please provide some more context to this scenario, ideally a standalone repro project that shows the issue. At the minimum please provide listing for DbContext, how GetMapping() gets incorporated into the query, and the query itself.

**Comment by ajcvickers** [comment]
**EF Team Triage:** Closing this issue as the requested additional details have not been provided and we have been unable to reproduce it.

*BTW this is a canned response and may have info or details that do not directly apply to this particular issue. While we'd like to spend the time to uniquely address every incoming issue, we get a lot traffic on the EF projects and that is not practical. To ensure we maximize the time we have to work on fixing bugs, implementing new features, etc. we use canned responses for common triage decisions.*
